### PR TITLE
add kubelet configflags:experimental-pleg-relist-threshold

### LINKED ISSUE #45419: Node flapping between Ready/NotReady with PLEG issues
**Is this a request for help?** No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): PLEG NotReady kubelet

---

**Is this a BUG REPORT or FEATURE REQUEST?** Bug

**Kubernetes version** (use `kubectl version`): 1.6.2

**Environment**:
- **Cloud provider or hardware configuration**: CoreOS on AWS
- **OS** (e.g. from /etc/os-release):CoreOS 1353.7.0
- **Kernel** (e.g. `uname -a`): 4.9.24-coreos
- **Install tools**:
- **Others**:

**What happened**:

I have a 3-worker cluster. Two and sometimes all three nodes keep dropping into `NotReady`with the following messages in `journalctl -u kubelet`:

```
May 05 13:59:56 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 13:59:56.872880    2858 kubelet_node_status.go:379] Recording NodeNotReady event message for node ip-10-50-20-208.ec2.internal
May 05 13:59:56 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 13:59:56.872908    2858 kubelet_node_status.go:682] Node became not ready: {Type:Ready Status:False LastHeartbeatTime:2017-05-05 13:59:56.872865742 +0000 UTC LastTransitionTime:2017-05-05 13:59:56.872865742 +0000 UTC Reason:KubeletNotReady Message:PLEG is not healthy: pleg was last seen active 3m7.629592089s ago; threshold is 3m0s}
May 05 14:07:57 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:07:57.598132    2858 kubelet_node_status.go:379] Recording NodeNotReady event message for node ip-10-50-20-208.ec2.internal
May 05 14:07:57 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:07:57.598162    2858 kubelet_node_status.go:682] Node became not ready: {Type:Ready Status:False LastHeartbeatTime:2017-05-05 14:07:57.598117026 +0000 UTC LastTransitionTime:2017-05-05 14:07:57.598117026 +0000 UTC Reason:KubeletNotReady Message:PLEG is not healthy: pleg was last seen active 3m7.346983738s ago; threshold is 3m0s}
May 05 14:17:58 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:17:58.536101    2858 kubelet_node_status.go:379] Recording NodeNotReady event message for node ip-10-50-20-208.ec2.internal
May 05 14:17:58 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:17:58.536134    2858 kubelet_node_status.go:682] Node became not ready: {Type:Ready Status:False LastHeartbeatTime:2017-05-05 14:17:58.536086605 +0000 UTC LastTransitionTime:2017-05-05 14:17:58.536086605 +0000 UTC Reason:KubeletNotReady Message:PLEG is not healthy: pleg was last seen active 3m7.275467289s ago; threshold is 3m0s}
May 05 14:29:59 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:29:59.648922    2858 kubelet_node_status.go:379] Recording NodeNotReady event message for node ip-10-50-20-208.ec2.internal
May 05 14:29:59 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:29:59.648952    2858 kubelet_node_status.go:682] Node became not ready: {Type:Ready Status:False LastHeartbeatTime:2017-05-05 14:29:59.648910669 +0000 UTC LastTransitionTime:2017-05-05 14:29:59.648910669 +0000 UTC Reason:KubeletNotReady Message:PLEG is not healthy: pleg was last seen active 3m7.377520804s ago; threshold is 3m0s}
May 05 14:44:00 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:44:00.938266    2858 kubelet_node_status.go:379] Recording NodeNotReady event message for node ip-10-50-20-208.ec2.internal
May 05 14:44:00 ip-10-50-20-208.ec2.internal kubelet[2858]: I0505 14:44:00.938297    2858 kubelet_node_status.go:682] Node became not ready: {Type:Ready Status:False LastHeartbeatTime:2017-05-05 14:44:00.938251338 +0000 UTC LastTransitionTime:2017-05-05 14:44:00.938251338 +0000 UTC Reason:KubeletNotReady Message:PLEG is not healthy: pleg was last seen active 3m7.654775919s ago; threshold is 3m0s}
```

docker daemon is fine (local `docker ps`, `docker images`, etc. all work and respond immediately). 

using weave networking installed via `kubectl apply -f https://git.io/weave-kube-1.6`

**What you expected to happen**:

Nodes to be ready.

**How to reproduce it** (as minimally and precisely as possible):

Wish I knew how!

**Anything else we need to know**:

All of the nodes (workers and masters) on same private subnet with NAT gateway to Internet. Workers in security group that allows unlimited access (all ports) from masters security group; masters allow all ports from same subnet. proxy is running on workers; apiserver, controller-manager, scheduler on masters. 

`kubectl logs` and `kubectl exec` always hang, even when run from the master itself (or from outside).

### PR DESCRIPTION
**Pull request by linyouchong**
**What this PR does / why we need it**:
add kubelet configflags:experimental-pleg-relist-threshold
Kubelet will reports PLEG error and be notReady if pleg relist take times greater than experimental-pleg-relist-threshold. If we add this flag user can set a suitable threshold to avoid flipping between healthy and unhealthy. The default value is '3m'.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #45419

**Special notes for your reviewer**:
NONE

**Release note**:

```kubelet: `--experimental-pleg-relist-threshold` can now be specified pleg-relist-threshold, Kubelet will reports PLEG error and be notReady if pleg relist take times greater than experimental-pleg-relist-threshold.

```

### CODE REVIEW

**Comment by jianglingxia** [comment]
/ok-to-test
maybe you can remove the content in Release note to the place of What this PR does / why we need it,and then write NONE in the Release note,thanks!

**Comment by jianglingxia** [comment]
PTAL,thanks
/assign @random-liu

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -400,6 +400,9 @@ func (f *KubeletFlags) AddFlags(fs *pflag.FlagSet) {
 
 // AddKubeletConfigFlags adds flags for a specific kubeletconfig.KubeletConfiguration to the specified FlagSet
 func AddKubeletConfigFlags(fs *pflag.FlagSet, c *kubeletconfig.KubeletConfiguration) {
+
+	fs.DurationVar(&c.PlegRelistThreshold.Duration, "experimental-pleg-relist-threshold", c.PlegRelistThreshold.Duration, "Kubelet will reports PLEG error and be notReady if pleg relist take times greater than experimental-pleg-relist-threshold. Needs to be greater than the pleg relisting period(1s) + the relisting time, which can vary significantly. Set a conservative threshold to avoid flipping between healthy and unhealthy. Default: '3m'")
```
Comment (tianshapjq) [inline cmd/kubelet/app/options/options.go]: shorten this would be better

File: pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go [hunk 2]
```diff
@@ -246,6 +246,8 @@ type KubeletConfiguration struct {
 	FeatureGates map[string]bool `json:"featureGates,omitempty"`
 	// Tells the Kubelet to fail to start if swap is enabled on the node.
 	FailSwapOn *bool `json:"failSwapOn,omitempty"`
+	// Kubelet will reports PLEG error and be notReady if pleg relist take times greater than experimental-pleg-relist-threshold.
```
Comment (tianshapjq) [inline pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go]: reports -> report
take times -> takes time

File: pkg/kubelet/kubelet_test.go [hunk 3]
```diff
@@ -269,7 +269,7 @@ func newTestKubeletWithImageList(
 	kubelet.resyncInterval = 10 * time.Second
 	kubelet.workQueue = queue.NewBasicWorkQueue(fakeClock)
 	// Relist period does not affect the tests.
-	kubelet.pleg = pleg.NewGenericPLEG(fakeRuntime, 100, time.Hour, nil, clock.RealClock{})
+	kubelet.pleg = pleg.NewGenericPLEG(fakeRuntime, 100, time.Hour, 3*time.Minute, nil, clock.RealClock{})
```
Comment (tianshapjq) [inline pkg/kubelet/kubelet_test.go]: how about another test case on other thresholds? 3 is just default and may not be enough.

File: pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go (hunk 2 above)
Comment (linyouchong) [reply pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go]: Fixed, thanks for your review.

File: pkg/kubelet/kubelet_test.go (hunk 3 above)
Comment (linyouchong) [reply pkg/kubelet/kubelet_test.go]: there is one at pkg/kubelet/pleg/generic_test.go#TestHealthy

**Pull request closed without merging**